---
#############################
#  Input Module Parameters  #
#############################
inputs:
  - ns: input
    sensor:
      type: camera
      width: 640
      height: 360
      cx: 315.9902073557914
      cy: 254.59437002701452
      fx: 372.4634094238281
      fy: 371.5072021484375
      min_range: $(arg sensor_min_range)
      max_range: $(arg sensor_max_range)
      extrinsics:
        type: identity
input:
  global_frame_name: $(arg odom_frame)
  sensor_frame_name: $(arg sensor_frame)
  subscriber:
    type: $(arg subscriber_type)
    input_images_are_const: false
    color_image_topic: color/image_raw
    depth_image_topic: depth/image_rect
    semantic_image_topic: semantic/image_raw
    queue_size: 1000
#############################################
#  Active-Window Reconstruction Parameters  #
#############################################
reconstruction:
  use_visualizer: true
  active_window_viz:
    global_frame_name: $(arg odom_frame)
    color_mode: color
    slice_height: -0.5
    slice_height_is_relative: true
    activity_shading_intensity: 0.6
    object_color_mode: 0
    dynamic_point_scale: 0.1
  active_window:  # taken from Khronos
    verbosity: 1
    volumetric_map:
      voxel_size: 0.15  # m
      truncation_distance: 0.35  # m (Usually 2-3x voxel size)
      voxels_per_side: 16
    object_detector:
      type: $(arg object_detection_type)
      max_range: $(arg sensor_max_range)
    tracker:
      type: max_iou   # 'max_iou'
      track_by: voxels   # 'pixels', 'voxels', 'bounding_box'
    projective_integrator:
      max_weight: 100000
      interpolation_method: adaptive   # "nearest", "bilinear", "adaptive"
      verbosity: 1
    tracking_integrator:
      temporal_window: $(arg temporal_window_size_s)
      verbosity: 1
#########################
#  Frontend Parameters  #
#########################
frontend:
  type: LLMFrontend
  pgmo:
    horizon: 60.0
    d_graph_resolution: 2.5
    output_mesh_resolution: 0.02
  min_object_vertices: 0
  prune_mesh_indices: true
  places:
    type: gvd
    gvd:
      max_distance_m: 3.0
      min_diff_m: 0.1
      min_basis_for_extraction: 1
      voronoi_config:
        min_distance_m: 0.4
        parent_l1_separation: 25
        parent_cos_angle_separation: 0.1
    graph:
      type: CompressionGraphExtractor
      compression_distance_m: 1.0
      min_node_distance_m: 0.4
      min_edge_distance_m: 0.25
      node_merge_distance_m: 0.5
      add_heuristic_edges: true
      add_overlap_edges: false
      add_freespace_edges: true
      freespace_edges:
        min_clearance_m: 0.3
        num_nodes_to_check: 10
        num_neighbors_to_find: 3
#    sinks:
      #- type: PlacesVisualizer
        # ns: ~sink_configs/places
  view_database:
    view_selection_method: $(arg place_view_selection_type)
########################
#  Backend Parameters  #
########################
backend:
  type: BackendModule
  dsg:
    places_merge_distance_tolerance_m: 0.4
    enable_node_merging: true
  pgmo:
    run_mode: 0  # kimera_pgmo run mode FULL (required)
    use_msg_time: true
    embed_trajectory_delta_t: 5.0
    num_interp_pts: 3
    interp_horizon: 10.0
    enable_sparsify: false
    trans_node_dist: 1.0
    rot_node_dist: 1.2
    rpgo:
      odom_trans_threshold: 0.05
      odom_rot_threshold: 0.01
      pcm_trans_threshold: -1
      pcm_rot_threshold: -1
      gnc_alpha: 0.9
      gnc_mu_step: 1.6
      gnc_fix_prev_inliers: true
      verbosity: UPDATE
      solver: LM
    covariance:
      odom: 1.0e-02
      loop_close: 5.0e-02
      sg_loop_close: 1.0e-01
      prior: 1.0e-02
      mesh_mesh: 1.0e-02
      pose_mesh: 1.0e-02
      place_mesh: 1.0e-02
      place_edge: 10.0
      place_merge: 10.0
      object_merge: 10.0
  enable_rooms: false
  objects:
    tasks:
      type: RosEmbeddingGroup
  regions:
    color_by_task: true
    extractor:
      type: $(arg places_clustering_type)
      clustering:
        type: AgglomerativeClustering  # unused if doing geometric rooms
        tasks:
          type: RosEmbeddingGroup
          prompts: $(arg places_task_prompts)
          silent_wait: false
        metric:
          type: cosine
        selector:
          type: IB
          max_delta: 0.015
          score_threshold: 0.21
          top_k: 2
          cumulative: true
          null_task_preprune: true
      rooms:
        min_dilation_m: 0.5
        max_dilation_m: 1.2
        min_window_size: 0.2
        clip_dilation_window_to_max: false
        min_component_size: 10
        min_room_size: 10
        dilation_threshold_mode: PLATEAU
        min_lifetime_length_m: 0.1
        plateau_ratio: 0.25
        clustering_mode: NEIGHBORS
        max_modularity_iters: 5
        modularity_gamma: 1.0
        room_prefix: R
        dilation_diff_threshold_m: -1.0
        log_filtrations: false
        log_place_graphs: false
sink_configs:
  places:
    gvd_visualizer:
      visualization_type: 2  # 0 -> NONE, 1 -> ESDF_SLICE, 2 -> GVD
      color_nearest_vertices: false
      gvd_alpha: 0.6
      gvd_min_distance: 0.0
      gvd_max_distance: 1.0
      basis_threshold: 2
      min_num_basis: 1
      max_num_basis: 10
      gvd_mode: 0  # 0 -> DEFAULT, 1 -> DISTANCE, 2 -> BASIS_POINTs
      esdf_alpha: 0.6
      slice_height: 2.0
      esdf_min_distance: 0.0
      esdf_max_distance: 2.5
    graph_visualizer:
      z_offset_scale: 0.0
      visualize: true
      marker_scale: 0.2
      marker_alpha: 0.9
      use_sphere_marker: true
      use_label: true
      label_height: 0.7
      label_scale: 0.3
      use_bounding_box: false
      bounding_box_alpha: 0.5
      use_edge_source: false
      interlayer_edge_scale: 0.1
      interlayer_edge_alpha: 0.4
      interlayer_edge_use_color: true
      interlayer_edge_insertion_skip: 0
      intralayer_edge_scale: 0.02
      intralayer_edge_alpha: 0.4
      intralayer_edge_use_color: false
      intralayer_edge_insertion_skip: 0
    visualizer_colormap:
      min_hue: 0.0
      max_hue: 0.5
      min_saturation: 0.9
      max_saturation: 0.8
      min_luminance: 0.6
      max_luminance: 0.85
